lib/OperationHandler.cpp

```diff
@@ -64,87 +62,139 @@ void OperationHandler::SendGuestNotification(std::variant<DisconnectNotif, Signa
+            ToString(status),
+            GuidToString(interfaceGuid).c_str(),
+            SsidToLogString(ssid.value()).c_str(),
+            Wlansvc::AuthAlgoToString(authAlgo).c_str());
+
+        if (m_clientObserver)
+        {
+            m_clientObserver->OnGuestConnectionCompletion(type, status, {interfaceGuid, ssid, authAlgo});
+        }
+    });
+}
 
-    if (m_clientCallbacks.OnDisconnection)
-    {
-        m_clientCallbacks.OnDisconnection(source, {interfaceGuid, network});
-    }
+void OperationHandler::OnGuestDisconnectionRequest(OperationType type, const Ssid& ssid) noexcept
+{
+    m_clientNotificationQueue.RunAndWait([&] {
+        Log::Info(
+            L"Notifying the client of a guest disconnection request. Type: %ws, Ssid: %ws",
+            ToString(type),
+            SsidToLogString(ssid.value()).c_str());
+
+        if (m_clientObserver)
+        {
+            m_clientObserver->OnGuestDisconnectionRequest(type, ssid);
+        }
+    });
 }
 
-void OperationHandler::NotifyConnectionToClient(EventSource source, const GUID& interfaceGuid, const DOT11_SSID& network, DOT11_AUTH_ALGORITHM authAlgo)
+void OperationHandler::OnGuestDisconnectionCompletion(OperationType type, OperationStatus status, const GUID& interfaceGuid, const Ssid& ssid) noexcept
 {
-    m_clientNotificationQueue.RunAndWait([this, source, interfaceGuid, network, authAlgo] {
-        NotifyConnectionToClientSerialized(source, interfaceGuid, network, authAlgo);
+    m_clientNotificationQueue.RunAndWait([&] {
+        Log::Info(
+            L"Notifying the client of a guest disconnection completion. Type: %ws, Status: %ws, Interface: %ws, Ssid: %ws",
+            ToString(type),
+            ToString(status),
+            GuidToString(interfaceGuid).c_str(),
+            SsidToLogString(ssid.value()).c_str());
+
+        if (m_clientObserver)
+        {
+            m_clientObserver->OnGuestDisconnectionCompletion(type, status, {interfaceGuid, ssid});
+        }
     });
 }
 
-void OperationHandler::NotifyDisconnectionToClient(EventSource source, const GUID& interfaceGuid, const DOT11_SSID& network)
+void OperationHandler::OnGuestScanRequest() noexcept
 {
-    m_clientNotificationQueue.RunAndWait(
-        [this, source, interfaceGuid, network] { NotifyDisconnectionToClientSerialized(source, interfaceGuid, network); });
+    m_clientNotificationQueue.RunAndWait([&] {
+        Log::Info(L"Notifying the client of a guest scan request");
+
+        if (m_clientObserver)
+        {
+            m_clientObserver->OnGuestScanRequest();
+        }
+    });
 }
 
-void OperationHandler::NotifyGuestConnectRequestProgress(GuestConnectStatus status)
+void OperationHandler::OnGuestScanCompletion(OperationStatus status) noexcept
 {
-    m_clientNotificationQueue.RunAndWait([this, status] {
-        Log::Info(L"Notifying guest directed connection progress to the client. Status: %ws", GuestConnectStatusToString(status));
+    m_clientNotificationQueue.RunAndWait([&] {
```

Comment:
Why the capture mode change here? It seemed like capturing `status` by value was the right thing to do.

Reply:
Since we wait for the task completion, the variable won't go out of scope and a reference is enough.

Reply:
Thanks, the `RunAndWait` part escaped me.